### ISSUE TITLE
NullReferenceExceptions - audit code for potential null derefs

### ISSUE THREAD

**Issue by Scottj1s** [microsoft/CsWinRT#514]
With some recent NREs on string compares, a comprehensive check for nulls in the code would be beneficial.

Some techniques to automate, vs a manual scan of the code:

Ensure that fxcop is flagging public strings via CA1062:
https://docs.microsoft.com/en-ca/dotnet/fundamentals/code-analysis/quality-rules/ca1062?view=vs-2019

Run a few Roslyn analyzers:
http://pihrt.net/Roslynator/Analyzers

Enable (really, disable) nullable references:
https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/

The last option is the most comprehensive and would detect any unguarded nullables (not just strings), which may induce a bit of churn, but would probably be worthwhile.

**Comment by AdamBraden** [comment]
Create a branch with this enabled and check in

**Comment by j0shuams** [comment]
TL;DR: Cost outweighs benefit. About 1k warnings to handle in WinRT.Runtime, and more to be discovered once the first wave of annotations are added (e.g. newly annotated functions are used in the return statement of other functions, which would also now need annotations). More changes would be needed as well, and there is not much to gain from enabling null references. 

Did due diligence and played around with nullable references enabled. Caused about 100k warnings to be generated, with about 1% of those living in WinRT.Runtime. Having gone in and added `?` to return types of ABI functions like `MarshalInspectable.FromAbi`, the functions that use the ABI functions will in-turn need `?`. So there will be more work than just the 1%, once changes start to be made. And this is just for the manually projected code, to get proper changes in for generated code we would need to update `code_writers.h` to maintain state/context about what type is being written so it would know whether to add a `?` to the end of a type name. So, there's a lot of work and not a lot of benefit.
